internal/jimm/role/role.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 Canonical.
+package role
+
+import (
+	"context"
+
+	"github.com/canonical/jimm/v3/internal/common/pagination"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/openfga"
+	jimmnames "github.com/canonical/jimm/v3/pkg/names"
+)
+
+// RoleStore defines a storage interface to persist roles.
+type RoleStore interface {
+	// AddRole adds a new role.
+	AddRole(ctx context.Context, name string) (re *dbmodel.RoleEntry, err error)
+
+	// GetRole populates the provided *dbmodel.RoleEntry based on name or UUID.
+	GetRole(ctx context.Context, role *dbmodel.RoleEntry) (err error)
+
+	// UpdateRoleName updates the name of a role identified by UUID.
+	UpdateRoleName(ctx context.Context, uuid, name string) (err error)
+
+	// RemoveRole removes the role identified by its ID or UUID.
+	RemoveRole(ctx context.Context, role *dbmodel.RoleEntry) (err error)
+
+	// ListRoles returns a paginated list of Roles defined by limit and offset.
+	// match is used to fuzzy find based on entries' name or uuid using the LIKE operator (ex. LIKE %<match>%).
+	ListRoles(ctx context.Context, limit, offset int, match string) (_ []dbmodel.RoleEntry, err error)
+
+	// CountRoles returns a count of the number of Roles that exist.
+	CountRoles(ctx context.Context) (count int, err error)
+}
+
+// RoleAuthorisationService holds the interface to manage roles within an authorisation service.
+type RoleAuthorisationService interface {
+	// RemoveRole removes a role.
+	RemoveRole(ctx context.Context, role jimmnames.RoleTag) error
+}
+
+// roleManager provides a means to manage roles within JIMM.
+type roleManager struct {
+	store   RoleStore
+	authSvc RoleAuthorisationService
+}
+
+// NewRoleManager returns a new RoleManager that persists the roles in the provided store.
+func NewRoleManager(store RoleStore, authSvc RoleAuthorisationService) (*roleManager, error) {
+	if store == nil {
+		return nil, errors.E("role store cannot be nil")
+	}
+	if authSvc == nil {
+		return nil, errors.E("role authorisation service cannot be nil")
+	}
+	return &roleManager{store, authSvc}, nil
+}
+
+// AddRole adds a role to JIMM.
+func (rm *roleManager) AddRole(ctx context.Context, user *openfga.User, roleName string) (*dbmodel.RoleEntry, error) {
```

Comment:
nit: i would prefer `rlM` just because `rm` down the code seems the UNIX `rm`

Reply:
Not sure if this really matters so I'll ignore this one if that's ok?

Reply:
yy, for sure. Just a thing I had on my mind